rainier-core/src/main/scala/com/stripe/rainier/sampler/LeapFrog.scala

```diff
@@ -60,6 +60,85 @@ private[sampler] case class LeapFrog(density: DensityFunction) {
     pqBuf(potentialIndex) = density.density * -1
   }
 
+  /**
+    * Perform l leapfrog steps starting at position q and momentum p
+    * @param l the total number of leapfrog steps to perform
+    * @return the new value of the parameters and momentum
+    * array with updated density
+    */
+  private def steps(l: Int, stepSize: Double): Unit = {
+    initialHalfThenFullStep(stepSize)
+    var i = 1
+    while (i < l) {
+      twoFullSteps(stepSize)
+      i += 1
+    }
+    finalHalfStep(stepSize)
+  }
+
+  private def isUTurn(theta: Array[Double], pqP: Array[Double]): Boolean = {
+
+    var out = 0.0
+    var i = 0
+    while (i < theta.size) {
+      out += (pqP(i + nVars) - theta(i)) * pqP(i)
+      i += 1
+    }
+
+    if (out.isNaN)
+      true
+    else
+      out < 0
+  }
+
+  /**
+    * Calculate the longest-step size until a u-turn
+    */
+  private def longestStepUntilUTurn(l0: Int, stepSize: Double): Int = {
+
+    val initTheta = variables(pqBuf)
+    var out = pqBuf
+    var l = 0
+    while (!isUTurn(initTheta, pqBuf)) {
+      l += 1
+      steps(1, stepSize)
+
+      if (l == l0)
+        out = pqBuf
```

Comment:
This is just aliasing a mutable array. If you want to be snapshotting the state at the `l0`th step, so that you can rewind to it after keeping going until a u-turn, you need to be using `copy` into some other buffer. I suggest you allocate the buffer once in `empiricalLongestSteps` and thread it through to here.